posts/spring2021/js-chats-7-typescript/index.md

```diff
@@ -0,0 +1,415 @@
+
+Lets say we are interacting with a web API that returns responses of this shape:
+
+```tsx
+interface ApiResponse {
+	status: boolean;
+	message: string;
+	data: any;
+}
+```
+
+Although giving the response data a type of *any* saves us from typing every API endpoint, this opens our application up to type-safety errors, as TypeScript will not type-check the data property.
+
+To work around this, we can use generics! We can specify a template *T* scoped to our interface and pass it to our dynamic property *data*.
+
+```tsx
+interface ApiResponse<T> {
+	status: boolean;
+	message: string;
+	data: T;
+}
+type PersonApiResponse = ApiResponse<Person>;
+```
+
+Let's explore another use case of generics is in the context of functions.
+
+In Python, getting the last element of a list is as simple as getting the element at index -1, however, the same functionality in JavaScript involves retrieving the array's length at runtime and normalizing it to be zero-indexed.
+
+```tsx
+const lastElement = elements[elements.length - 1]
+```
+
+Let's create a helper function *lastOf* that does this and returns a value of a type based on the input array. We can do this by creating a template T scoped to our *lastOf* function that is then used to describe the shape of the input and output types.
+
+```tsx
+function lastOf<T>(elements: T[]): T {
+	return elements[elements.length - 1];
+}
+
+// Explicitly passing generics
+let num = lastOf<number>([1, 6, 3, 5]);
+
+// TypeScript infers that this variable is a string.
+let str = lastOf(["an", "array", "of", "strings"]);
+
+str = 2 // TypeScript error, str is a string
+num = "Not a number" // TypeScript error, num is a number
+```
+
+## Library Support
+
+npm libraries largely fall under three categories:
+
+1. Packages with types defined internally
+2. Popular packages without defined types, but with community supplied types
+3. Not-so-popular packages without defined types or community supplied types
+
+With the first class of npm packages, no changes need to be made. However, for the second class of packages, there is a [community-supported monorepo](https://github.com/DefinitelyTyped/DefinitelyTyped) from which types can be seamlessly added. For instance, to install the types for *lodash* (a package that doesn't provide types) as a development dependency, we can run the following command:
+
+```tsx
+npm install --save-dev @types/lodash
+```
+
+All types and TypeScript related packages should be installed as development dependencies, as they only used in development and not in production builds.
+
+For the third class of packages, we can define types based on the package's documentation. [Here](https://www.credera.com/insights/typescript-adding-custom-type-definitions-for-existing-libraries)'s a useful resource for those looking to write custom types for a package.
```

Comment:
nit: It's usually better to include some information on what the link is in the prose itself:

```suggestion
For the third class of packages, we can define types based on the package's documentation. [Credera](https://www.credera.com/insights/typescript-adding-custom-type-definitions-for-existing-libraries) has some useful resources for those looking to write custom types for a package.
```